### ISSUE TITLE
Wrong signature of arguments for rule with function wrapper?

### ISSUE THREAD

**Issue by gdalle** [compintell/Mooncake.jl#273]
A version of this occurred when I tried to benchmark Mooncake with DifferentiationInterfaceTest.

```julia
using DifferentiationInterface
import Mooncake

backend = AutoMooncake(; config=nothing)

struct Caller{F}
    f::F
end

(c::Caller)(args...) = c.f(args...)

Caller(sum)([1.0])
gradient(sum, backend, [1.0])  # works
gradient(Caller(sum), backend, [1.0])  # fails
```

The stacktrace reads

```julia
julia> gradient(Caller(sum), backend, [1.0])  # fails
ERROR: ArgumentError: signature of arguments, Tuple{Mooncake.CoDual{Caller{typeof(sum)}, Mooncake.NoFData}, Mooncake.CoDual{Vector{Float64}, Vector{Float64}}}, not equal to signature required by rule, Tuple{Mooncake.CoDual{Caller{typeof(sum)}, Mooncake.NoFData}, Mooncake.CoDual{Tuple{Vector{Float64}}, Tuple{Vector{Float64}}}}.
Stacktrace:
 [1] __verify_sig(::Mooncake.DerivedRule{…}, ::Tuple{…})
   @ Mooncake ~/.julia/packages/Mooncake/tUYT3/src/interface.jl:26
 [2] __value_and_pullback!!(::Mooncake.DerivedRule{…}, ::Float64, ::Mooncake.CoDual{…}, ::Mooncake.CoDual{…})
   @ Mooncake ~/.julia/packages/Mooncake/tUYT3/src/interface.jl:13
 [3] value_and_pullback!!(::Mooncake.DerivedRule{…}, ::Float64, ::Caller{…}, ::Vector{…})
   @ Mooncake ~/.julia/packages/Mooncake/tUYT3/src/interface.jl:116
 [4] value_and_pullback(::Caller{…}, ::DifferentiationInterfaceMooncakeExt.MooncakeOneArgPullbackPrep{…}, ::AutoMooncake{…}, ::Vector{…}, ::Tuple{…})
   @ DifferentiationInterfaceMooncakeExt ~/Work/GitHub/Julia/DifferentiationInterface.jl/DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/onearg.jl:32
 [5] prepare_pullback(::Caller{typeof(sum)}, ::AutoMooncake{Nothing}, ::Vector{Float64}, ::Tuple{Bool})
   @ DifferentiationInterfaceMooncakeExt ~/Work/GitHub/Julia/DifferentiationInterface.jl/DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/onearg.jl:18
 [6] prepare_gradient(::Caller{typeof(sum)}, ::AutoMooncake{Nothing}, ::Vector{Float64})
   @ DifferentiationInterface ~/Work/GitHub/Julia/DifferentiationInterface.jl/DifferentiationInterface/src/first_order/gradient.jl:58
 [7] gradient(::Caller{typeof(sum)}, ::AutoMooncake{Nothing}, ::Vector{Float64})
   @ DifferentiationInterface ~/Work/GitHub/Julia/DifferentiationInterface.jl/DifferentiationInterface/src/fallbacks/no_prep.jl:11
 [8] top-level scope
   @ ~/Work/GitHub/Julia/DifferentiationInterface.jl/DifferentiationInterfaceTest/test/playground.jl:14
Some type information was truncated. Use `show(err)` to see complete types.
```

**Comment by willtebbutt** [comment]
Ah, this is annoying. I need to add some tests for `value_and_pullback!!` with varargs functions.

Does a version of this also happen with the two-arg interface, or do we not have this probelm there?

**Comment by willtebbutt** [comment]
@gdalle could you confirm that this problem is resolved if you use version 0.4.5, and close this issue if so?

**Comment by gdalle** [comment]
Thanks, this is now fixed for both out-of-place and in-place functions
